<img src="img/Marca-ITBA-Color-ALTA.png" width="250">

# Master en Management & Analytics

## Programación para el Análisis de Datos

## Clase 3 - Práctica resuelta

In [ ]:
import numpy as np
import pandas as pd
print(pd.__version__)

df = pd.read_csv('data/phone_data.csv', index_col=0)
df.head()

1.4.4


date  duration  item    month   network network_type
index                                                                
0      15/10/14 06:58    34.429  data  2014-11      data         data
1      15/10/14 06:58    13.000  call  2014-11  Vodafone       mobile
2      15/10/14 14:46    23.000  call  2014-11    Meteor       mobile
3      15/10/14 14:48     4.000  call  2014-11     Tesco       mobile
4      15/10/14 17:27     4.000  call  2014-11     Tesco       mobile

### Ejercicio 1

Particionar el `Dataframe` por `network_type`. Mostrar por pantalla unicamente el grupo `'network_type'='data'`

**Tips de funciones a utilizar:**
  - groupby
  - get_group

In [ ]:
### Completa con tu código aquí
df_grouped = df.groupby("network_type")
df_grouped.groups['data']

Int64Index([  0,   6,  13,  26,  39,  45,  54,  56,  58,  66,
            ...
            811, 817, 818, 819, 820, 821, 822, 823, 824, 827],
           dtype='int64', name='index', length=150)

### Ejercicio 2

Obtener el registro con duración mas larga para el grupo `'landline'`

**Tips de funciones a utilizar:**
  - groupby
  - get_group
  - sort_values

In [ ]:
#output_esperado =  10528.0


### Completa con tu código aquí
maxDurLandline = df_grouped.groups['landline'].max()
print(maxDurLandline)
display(df_grouped.get_group('landline').sort_values('duration').tail(1))
###


816


date  duration  item    month   network network_type  index
index                                                                       
816    04/03/15 12:29   10528.0  call  2015-03  landline     landline    816

### Ejercicio 3

Obtener la duración media de los eventos agrupados por `'month'` y`'item'`. 

Obtener, además, la última fecha de evento.

**Tips de funciones a utilizar:**
  - groupby
  - aggregate / agg
  - np.mean
  - last
  
  
output_esperado:

|                     |   duration | date           |
|:--------------------|-----------:|:---------------|
| ('2014-11', 'call') |    238.757 | 12/11/14 19:01 |
| ('2014-11', 'data') |     34.429 | 12/11/14 06:58 |
| ('2014-11', 'sms')  |      1     | 13/11/14 22:31 |
| ('2014-12', 'call') |    171.658 | 14/12/14 19:54 |
| ('2014-12', 'data') |     34.429 | 12/12/14 06:58 |
| ('2014-12', 'sms')  |      1     | 07/12/14 23:22 |
| ('2015-01', 'call') |    193.977 | 14/01/15 20:47 |
| ('2015-01', 'data') |     34.429 | 12/01/15 06:58 |
| ('2015-01', 'sms')  |      1     | 14/01/15 23:36 |
| ('2015-02', 'call') |    215.164 | 09/02/15 17:54 |
| ('2015-02', 'data') |     34.429 | 12/02/15 06:58 |
| ('2015-02', 'sms')  |      1     | 10/02/15 21:40 |
| ('2015-03', 'call') |    462.277 | 04/03/15 12:29 |
| ('2015-03', 'data') |     34.429 | 13/03/15 06:58 |
| ('2015-03', 'sms')  |      1     | 14/03/15 00:16 |

In [ ]:
### Completa con tu código aquí
pd.set_option('display.float_format', '{:,.3f}'.format)
df_groupedMI = df.groupby(['month', 'item'])
df_groupedMI.agg({'duration': 'mean',
                  'date': 'last'})
### Ver por qué la salida no me da igual. month e item no pertenecen al df de salida (por qué si aparecen?)

duration            date
month   item                          
2014-11 call   238.757  12/11/14 19:01
        data    34.429  12/11/14 06:58
        sms      1.000  13/11/14 22:31
2014-12 call   171.658  14/12/14 19:54
        data    34.429  12/12/14 06:58
        sms      1.000  07/12/14 23:22
2015-01 call   193.977  14/01/15 20:47
        data    34.429  12/01/15 06:58
        sms      1.000  14/01/15 23:36
2015-02 call   215.164  09/02/15 17:54
        data    34.429  12/02/15 06:58
        sms      1.000  10/02/15 21:40
2015-03 call   462.277  04/03/15 12:29
        data    34.429  13/03/15 06:58
        sms      1.000  14/03/15 00:16

### Ejercicio 4

Obtener los primeros `n=2` elementos de cada grupo considerando la columna `network_type` como clave para agrupar.

**Tips de funciones a utilizar:**
  - groupby
  - head
  
Output esperado:

|   index | date           |   duration | item   | month   | network   | network_type   |
|--------:|:---------------|-----------:|:-------|:--------|:----------|:---------------|
|       0 | 15/10/14 06:58 |     34.429 | data   | 2014-11 | data      | data           |
|       6 | 16/10/14 06:58 |     34.429 | data   | 2014-11 | data      | data           |
|      49 | 20/10/14 13:56 |      6     | call   | 2014-11 | landline  | landline       |
|      59 | 23/10/14 08:34 |   1940     | call   | 2014-11 | landline  | landline       |
|       1 | 15/10/14 06:58 |     13     | call   | 2014-11 | Vodafone  | mobile         |
|       2 | 15/10/14 14:46 |     23     | call   | 2014-11 | Meteor    | mobile         |
|     121 | 03/11/14 08:40 |      1     | sms    | 2014-11 | special   | special        |
|     596 | 15/01/15 12:23 |      1     | sms    | 2015-02 | special   | special        |
|       9 | 16/10/14 15:30 |     19     | call   | 2014-11 | voicemail | voicemail      |
|      32 | 18/10/14 14:01 |     96     | call   | 2014-11 | voicemail | voicemail      |
|     360 | 06/12/14 18:27 |      1     | sms    | 2014-12 | world     | world          |
|     361 | 06/12/14 18:28 |      1     | sms    | 2014-12 | world     | world          |

In [ ]:
### Tu código aquí:
df_groupedNT = df.groupby('network_type')
df_groupedNT.head(2).sort_values('network_type')
###


date  duration  item    month    network network_type  index
index                                                                        
0      15/10/14 06:58    34.429  data  2014-11       data         data    816
6      16/10/14 06:58    34.429  data  2014-11       data         data    816
49     20/10/14 13:56     6.000  call  2014-11   landline     landline    816
59     23/10/14 08:34 1,940.000  call  2014-11   landline     landline    816
1      15/10/14 06:58    13.000  call  2014-11   Vodafone       mobile    816
2      15/10/14 14:46    23.000  call  2014-11     Meteor       mobile    816
121    03/11/14 08:40     1.000   sms  2014-11    special      special    816
596    15/01/15 12:23     1.000   sms  2015-02    special      special    816
9      16/10/14 15:30    19.000  call  2014-11  voicemail    voicemail    816
32     18/10/14 14:01    96.000  call  2014-11  voicemail    voicemail    816
360    06/12/14 18:27     1.000   sms  2014-12      world        world    816
361    06/12/14 18:28     1.000   sms  2014-12      world        world    816

### Ejercicio 5

Crear una función que descarte los ultimos `n=200` registros de cada mes de facturación.

**Tips de funciones a utilizar:**
  - groupby
  - tail
  - index
  
  
index_esperado = 
[0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10,  11,  12,  13,  14,  15,  16,  17,  
18,  19,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  381,  386,  389,  390,  391]

In [ ]:
df_group = df.groupby('month')
index_complemento = df_group.tail(200).index
index_esperado = np.setdiff1d(df.index, index_complemento)
display(df.iloc[index_esperado])

date  duration  item    month    network network_type  index
index                                                                        
0      15/10/14 06:58    34.429  data  2014-11       data         data    816
1      15/10/14 06:58    13.000  call  2014-11   Vodafone       mobile    816
2      15/10/14 14:46    23.000  call  2014-11     Meteor       mobile    816
3      15/10/14 14:48     4.000  call  2014-11      Tesco       mobile    816
4      15/10/14 17:27     4.000  call  2014-11      Tesco       mobile    816
5      15/10/14 18:55     4.000  call  2014-11      Tesco       mobile    816
6      16/10/14 06:58    34.429  data  2014-11       data         data    816
7      16/10/14 15:01   602.000  call  2014-11      Three       mobile    816
8      16/10/14 15:12 1,050.000  call  2014-11      Three       mobile    816
9      16/10/14 15:30    19.000  call  2014-11  voicemail    voicemail    816
10     16/10/14 16:21 1,183.000  call  2014-11      Three       mobile    816
11     16/10/14 22:18     1.000   sms  2014-11     Meteor       mobile    816
12     16/10/14 22:21     1.000   sms  2014-11     Meteor       mobile    816
13     17/10/14 06:58    34.429  data  2014-11       data         data    816
14     17/10/14 10:53     1.000   sms  2014-11      Tesco       mobile    816
15     17/10/14 11:19     1.000   sms  2014-11      Tesco       mobile    816
16     17/10/14 11:20     1.000   sms  2014-11     Meteor       mobile    816
17     17/10/14 17:22     1.000   sms  2014-11   Vodafone       mobile    816
18     17/10/14 17:23     1.000   sms  2014-11   Vodafone       mobile    816
19     17/10/14 17:26    92.000  call  2014-11      Three       mobile    816
20     17/10/14 17:29     4.000  call  2014-11   Vodafone       mobile    816
21     17/10/14 17:30   375.000  call  2014-11      Tesco       mobile    816
22     17/10/14 17:42     1.000   sms  2014-11   Vodafone       mobile    816
23     17/10/14 17:44     1.000   sms  2014-11   Vodafone       mobile    816
24     17/10/14 17:44     1.000   sms  2014-11   Vodafone       mobile    816
25     17/10/14 17:44     1.000   sms  2014-11   Vodafone       mobile    816
26     18/10/14 06:58    34.429  data  2014-11       data         data    816
27     18/10/14 11:51   783.000  call  2014-11      Tesco       mobile    816
28     18/10/14 12:06     4.000  call  2014-11   Vodafone       mobile    816
29     18/10/14 12:06     3.000  call  2014-11   Vodafone       mobile    816
381    13/12/14 06:58    34.429  data  2015-01       data         data    816
386    14/12/14 06:58    34.429  data  2015-01       data         data    816
389    15/12/14 06:58    34.429  data  2015-01       data         data    816
390    15/12/14 19:56     1.000   sms  2015-01      Three       mobile    816
391    15/12/14 19:58     1.000   sms  2015-01      Three       mobile    816

### Ejercicio 6

* Generar una columna que contenga la hora de las llamadas

* Discretizar la hora de modo tal que se mapeen los rangos horarios a las siguientes categorías:

    - 0-5am - night
    - 5-12 - morning
    - 12-20 - afternoon
    - 20 - 0 evening
    
    
* Hacer una tabla pivot que muestre la duración media de las llamadas en función de la franja horaria y del mes

* Hacer una tabla pivot que muestre el número de llamadas en función de la franja horaria y del mes


In [ ]:
### Completa con tu código aquí
def mapRange(hour):
    if hour > '00:00' and hour <= '05:00':
        return "night"
    elif hour > '05:00' and hour <= '12:00':
        return "morning"
    elif hour > '12:00' and hour <= '20:00':
        return "afternoon"
    else:
        return "evening"

df['hora_llamada'] = df['date'].str.split(" ", expand=True)[1]
df['rango_horario'] = df['hora_llamada'].map(lambda x: mapRange(x))
display(df.pivot_table(index={'rango_horario', 'month'}, values='duration', aggfunc='mean'))
display(df.pivot_table(index={'rango_horario', 'month'}, aggfunc='count', values='item'))


duration
month   rango_horario          
2014-11 afternoon       149.950
        evening          64.606
        morning          93.725
        night             1.286
2014-12 afternoon       142.845
        evening          96.100
        morning          46.270
        night            50.000
2015-01 afternoon       113.235
        evening          78.581
        morning          54.058
        night            71.000
2015-02 afternoon       177.653
        evening          42.273
        morning          43.544
        night             1.000
2015-03 afternoon       467.756
        evening          64.167
        morning          63.601
        night             1.000

item
month   rango_horario      
2014-11 afternoon       119
        evening          33
        morning          71
        night             7
2014-12 afternoon        71
        evening          10
        morning          70
        night             6
2015-01 afternoon       102
        evening          43
        morning          57
        night             3
2015-02 afternoon        72
        evening          11
        morning          52
        night             2
2015-03 afternoon        41
        evening          12
        morning          44
        night             4